### PR TITLE
Openl3 0.4.0 - Support for Tensorflow 2

### PR DESCRIPTION
**Pull request by beasteers**
Figured I would push this out while I'm waiting for something else to build.

Related PR containing updated models: https://github.com/marl/openl3/pull/61

Setup Changes
 - Openl3 now requires `tensorflow>=2.0.0` and installs it by default (there is no longer a separate GPU package)
 - Now requires `kapre>=0.3.5` - TODO: make sure we have the exact minimum kapre version - I remember checking git blame, but haven't tested anything
 - keras as a standalone package was removed from dependencies (we're using tf.keras)
 - travis.yml: removed python 2.7 & 3.5 and added 3.7 & 3.8 since tensorflow [only supports 3.6-3.8](https://www.tensorflow.org/install)
   - needed to install Cython first for python 3.8 in order to install skimage (`RuntimeError: Cython >= 0.23.4 is required to build scikit-image from git checkout`)

Doc Changes
 - Changed tensorflow dependency message to reflect updates
 - Added "Choosing an Audio Frontend (CPU / GPU)" section to tutorial.rst

Code Changes
 - core.py
   - added params: `get_audio_embedding(frontend='auto')`, `process_audio_file(frontend='auto')`, `process_video_file(audio_frontend='auto')`
   - Added function `preprocess_audio(y, sr, input_repr=None)` that encapsulates the librosa frontend (as well as preprocessing for the kapre frontend)
     - for librosa, you pass the `input_repr` and for kapre inputs, you leave `input_repr=None`
 - cli.py - added cli flag (`--audio-frontend`)
 - models.py
   - added param `load_audio_embedding_model(frontend='kapre')`
   - using new kapre composite layer helpers `get_stft_magnitude_layer`
   - disabled latest mag2db code and patched in the legacy version (`kapre_v0_1_4_magnitude_to_decibel`)
   - kapre is now technically an optional dependency (will only try to import if we try to load a model with kapre frontend)
     - we still install it with setup.py, but if someone wanted to, they could install everything manually without kapre and openl3 should still work for the librosa frontend

Test Changes
 - we now have separate regression data for kapre/librosa
 - added tests for frontend model following the existing model tests
 - converted some tests to use `pytest.mark.parameterize` to avoid doubling the length of the tests for testing frontends

Dev Util Changes
 - added `tests/generate_regression.py` which generates new regression data
 - added `tests/package_weights.py` which takes the weights files in the openl3 package folder and gzips them for git push
 - added `tests/migration/remove_layers.py` which lets us strip out the spectrogram (or any other) layers
 - `tests/migration/` has a few other analysis things/notebooks that were used early on in the frontend testing

Before merging:
 - double check dependency versions
 - are the pinned versions still valid? might need some help with this one
 - Change models download url in setup.py to main repo (currently it's pointing at my fork so I could test with travis)
 - should we integrate changes from https://github.com/marl/openl3/pull/55?
 - should we run the classifier comparison one more time right before merging as a safety check? idk

### CODE REVIEW

File: tests/package_weights.py
Comment (justinsalamon) [inline tests/package_weights.py:19]: I'm assuming this is for developers, not users? If a user upgrade from v0.3.1 to v0.4.0, will everything "just work"?

**Review by justinsalamon — CHANGES_REQUESTED** [review]
Really awesome work @beasteers !

I've made some comments to expand/clarify some parts of the docs.

Regarding the code, my main comments are:
* Not sure we want to keep the `auto` option for inferring the front-end. Adding a `frontend=kapre` kwarg to user-facing functions and putting the onus on them to ensure they match that parameter to the type of model they supply the function (when they provide a model) might be a safer option.
* We should avoid overloading parameters as much as possible. We should only accept `kapre` or `librosa`, not True/False. Similarly, `audio` should always expect an audio signal. If we want to support taking in a (mel)spectrogram, we should add a new optional kwarg `S`, following the API design of Librosa (see inline comments for link to example).

Thanks!

File: openl3/core.py
Comment (beasteers) [reply openl3/core.py:215]: I agree with this - it sounds like a better option

File: openl3/models.py
Comment (beasteers) [reply openl3/models.py:109]: I get what you mean and I'm not opposed to removing this. 

The reason that I said that it would make more sense is that it's a yes or no question of whether you want to have the frontend when you load the model. Originally I had the parameter as `include_frontend=True`, but I changed it to be more consistent.

I can remove the parameter aliases tho.

File: openl3/core.py
Comment (beasteers) [reply openl3/core.py:52]: fair point. I can take it out.

File: openl3/core.py
Comment (beasteers) [reply openl3/core.py:17]: Good catch! Tbh I thought that openl3 was already importing librosa which is why I left it at the top level.

And I just realized that it's not in setup.py but the reason that it runs fine is that kapre is dependent on librosa.

So:
 - I'll move the librosa import inside of the frontend functions
 - (though that won't make much of a difference since kapre imports it anyways)
 - do we want to add librosa as an explicit dependency to setup.py?

File: openl3/core.py
Comment (beasteers) [reply openl3/core.py:176]: You've asked about this is numerous places so I guess I'll just pick one spot to respond haha.

To answer your question above (just in case), if the frontend is set to auto, it will look at the model input shape. If it has 3 dimensions i.e. `(batch, channel, time)` then it takes PCM, but if it has 4 dimensions i.e. `(batch frequency, time, channel)` then it takes spectrograms. If there wasn't a clear distinction between the two input types then I would be much less eager to try to infer it.

So I'm going to push back a bit here. 

I really don't see a scenario where detecting the frontend would introduce bugs and one could argue that having to specify the same thing in two places could lead to more bugs/frustration because it's one more variable that they have to think about and keep track of.

Personally I don't like making APIs unnecessarily verbose and I don't think it would protect against someone wanting to use kapre and accidentally using librosa, for example.

For instance:
```python
def process_things(files, ..., frontend='librosa'):
    model = openl3.load_audio_embedding_model(..., frontend=frontend)
    for f in files:
        openl3.process_audio_file(f, model=model, frontend=frontend)
        # vs 
        openl3.process_audio_file(f, model=model)

```
I don't see how the first option is safer than the second. And if `process_audio_file` gets refactored to somewhere else, then they not only have to track the model, but they have to keep an explicit record of their frontend along side the model. Whereas if we can detect it, then if they want to switch from kapre to librosa, they only have to make the change in one place.

I do admit that I'm not in love with adding the additional `'auto'` value, but I think it's the most clear way to express it. An alternative could be `'infer'` which is a bit more clear about what's happening. 
(sidenote: usually when I want something to be calculated internally, I use `None` as a `null`/do-the-default-thing value, but I think that its falsey-ness would confuse things here.)

To propose a solution:
I will add some validation checks in `_validate_audio_frontend` so that if a user passes an explicit frontend that doesn't match the expected shape of the model, it will throw an error (basically if they pass `frontend=kapre` with a model that doesn't have a 3d input, then we throw an error, and same with librosa/4d).

So if someone wants to ensure that they are using the correct frontend it will throw an error if it's different, but otherwise we should just assume that the user knows what they are doing and assume that when they pass a librosa-compatible model, that they want to use a librosa frontend.

File: tests/migration/README.md
Comment (beasteers) [reply tests/migration/README.md:5]: I'm wondering if we need to keep these files in the repo. but i'll make that change for the time being

File: tests/package_weights.py
Comment (beasteers) [reply tests/package_weights.py:19]: Yes that was basically a note for Ho-Hsiang. Honestly we could remove the files in that directory. The only thing I think we need to keep in there is the `remove_layers.py`

File: tests/migration/check_layer_by_layer.py
Comment (beasteers) [reply tests/migration/check_layer_by_layer.py:16]: this isn't version specific actually - "current" refers to whatever openl3 version is installed in the environment. I clarified it

this was the early script that compared old kapre (0.1.4) to new kapre (0.3.5) so it was pre-"lets add librosa"

File: setup.py
Comment (beasteers) [reply setup.py:90]: I removed this as a setup.py dependency - but it's still in the travis.yml pre-install steps. maybe we can limit the skimage version to not need to build from git checkout. I'm not sure.

File: openl3/core.py
Comment (beasteers) [reply openl3/core.py:17]: Actually, I was thinking abt this earlier. 

We should have 
```python
import skimage
from moviepy.video.io.VideoFileClip import VideoFileClip
```
be optional dependencies as well - only import if you're using the image network.

And we also have a seemingly unused import that we can probably remove
```python
import sklearn.decomposition
```

File: docs/changes.rst
Comment (beasteers) [reply docs/changes.rst:9]: not sure how detailed we want to get here but there's definitely more to add to this

File: docs/installation.rst
Comment (beasteers) [reply docs/installation.rst:29]: that's a good idea. I'll add in the old message for earlier openl3 versions

File: setup.py
Comment (beasteers) [reply setup.py]: I changed it cuz my brain confused PR approved with merged so now I need to change it back until we merge the models repo.

File: openl3/core.py
Comment (beasteers) [reply openl3/core.py:176]: Actually one thing that I noticed - when using the librosa frontend, `input_repr` is confusing because on one hand, it is a required argument for `load_audio_embedding_model` (and the first argument as well) but it doesn't really do anything.

In order for librosa to consider your specified `input_repr`, you have to pass it to `get_audio_embedding` as well.

So to fix that, I'm going to throw an error if they don't specify `input_repr` with the librosa frontend. That's something that we can't detect from the model.

I just wish it wasn't a required argument to load the model because it's confusing.

File: docs/changes.rst
Comment (justinsalamon) [reply docs/changes.rst:9]: Well, right now it's `kapre>=`, so we need to add _some_ detail :) At the very least the lowest kapre version you've tested it against during development, right? You could add 2-3 more points summarizing over improvements you've made (e.g. streamlining some of the tests etc.) - any changes to the API, in particular, should be listed (e.g. the `frontend` kwarg).

File: docs/installation.rst
Comment (justinsalamon) [reply docs/installation.rst:29]: sounds good, tnx

File: setup.py
Comment (justinsalamon) [reply setup.py:90]: ok cool, yes I think it's fine to have it in travis, that's just for us (testing). Limiting the skimage version might cause issues downstream, perhaps leave as is for now and see how things shake out?

File: tests/migration/README.md
Comment (justinsalamon) [reply tests/migration/README.md:5]: Thanks. It's a good documentation of how the migration was done, so I'd vote to keep it :)

File: tests/migration/check_layer_by_layer.py
Comment (justinsalamon) [reply tests/migration/check_layer_by_layer.py:16]: gotcha, ok cool, thanks

File: tests/package_weights.py
Comment (justinsalamon) [reply tests/package_weights.py:19]: per my other comment I don't mind keeping these in as they document the migration nicely

File: setup.py
Comment (beasteers) [reply setup.py]: oops just pushed a change that took it out. I can add it back but I'll hold off for now just to avoid going back and forth haha

Comment (justinsalamon) [reply setup.py]: sorry if I wasn't clear, I meant - keep it in for travis (yml), leave it out of setup.py. I believe that's the current state of the code, so I don't think it needs any changes in that regard

File: openl3/core.py
Comment (beasteers) [reply openl3/core.py]: I think this would be great and maybe we can add it in later. I'm always for having a fallback that gives you full flexibility (so you could use any frontend you want), but I don't think I really feel like implementing this now lolol.

File: setup.py
Comment (justinsalamon) [reply setup.py:17]: The models PR has been approved already, right? :) Feel free to merge the models PR, then update this line back to marl, wait for the tests to pass, and merge this one too! :D

File: setup.py
Comment (justinsalamon) [inline setup.py:53]: yay!

**Review by justinsalamon — APPROVED** [review]
Really solid work @beasteers ! LGTM, let's moige it!

**Comment by beasteers** [comment]
@justinsalamon I don't have permissions to merge either of them so imma need u to merge the models first ;p

**Comment by justinsalamon** [comment]
doh! on it

**Comment by justinsalamon** [comment]
@beasteers model merged!

**Comment by beasteers** [comment]
@justinsalamon this looks all good too! whenever you're ready I think we're good to merge.

**Pull request merged**
